Use correct Poetry config when collecting Poetry projects

oranav opened the pull request:
**Description:**
When collecting Poetry projects for caching, a `**/poetry.lock` glob is used.  However, in order to process the Poetry configuration, the "poetry" command is run from the repo's root directory; this causes Poetry to return an invalid configuration when there is a Poetry project inside an inner directory.

Instead of running a single Poetry command, glob for the same pattern, and configure Poetry for every discovered project.

**Related issue:**
#446 

**Check list:**
- [ ] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

oranav commented:
I still need to test this locally, hence this is a draft.

oranav commented:
@IvanZosimov Sorry for the long delay! This is finally ready for review. All tests are passing successfully, and I'm using my own branch as an action in another project.

IvanZosimov commented:
> I fixed the failing main workflow on Windows. Good catch.
> 
> That being said, I don't understand why e2e-cache fails with Pipenv and Windows. I haven't changed anything related to it. Can it be failing randomly?

Hi, @oranav 👋 Thank you for this PR and sorry for the delay! The problem with e2e test related to pipenv on Windows is most probably caused by recent pipenv minor tag update, I hope that they will repair it. You also have 3 failing unit-tests in Main workflow, could take a look at them?

oranav commented:
It seems that I've hit an unrelated bug that happens sporadically, which causes tests to timeout in some cases. Take a look at [this action for example](https://github.com/actions/setup-python/actions/runs/3244002148/jobs/5319547229), which has failed for the same reason.

I've fixed it in [this commit](https://github.com/oranav/setup-python/commit/9578b65db127c898fb1dd4de1d68475e473de542). In short, the cache distributor tests were looking for dependency file with their default glob (`**/requirements.txt` for example), which made it search for the file in the **project's root** instead of in the tests data folder. I've changed that, and it reduced the tests time on my machine from ~7s to less than 3 seconds.

After we run the workflows again I expect the main workflow to pass.

oranav commented:
Hey @IvanZosimov, can you please take a look?

oranav commented:
I'm not sure why main workflow failed with Python 2.7 (which I haven't changed at all); I've seen that it failed in [other](https://github.com/actions/setup-python/actions/runs/3359996628/jobs/5570183786#step:7:10) [runs](https://github.com/actions/setup-python/actions/runs/3352973492/jobs/5555472793#step:7:16) as well, so I suspect it's unrelated to my PR.
I rebased my commits on top of `main`. Can we please re-run?

oranav commented:
@dsame @IvanZosimov Can we please re-run the actions?

oranav commented:
Sorry again. The problem was indeed unrelated to my PR, and it seems that #535 fixed the problem (for the curious - there is no Python 2.7 for Ubuntu 22.04, and the workflow previously used `ubuntu-latest` which got upgraded to 22.04). Unfortunately, #535 wasn't included in my rebase; I now rebased on top of it as well. Running again should result in a working workflow. Can we please re-run one more time?

I ran the workflow in my repo just in case, and it passed successfully:
https://github.com/oranav/setup-python/actions/runs/3412737221/jobs/5678633730

oranav commented:
@dsame @IvanZosimov Can we please re-run the workflow?
Sorry for the back and forth :)

dsame commented:
@oranav can you please explain the purpose of running `poetry env use` after the `.venv` already has been added to the list of cached directories? Is it just to create `.venv` directory? 

i mean this https://github.com/actions/setup-python/pull/447/files#diff-5a0a81991280dd0bbc75f051187afdb8ca7cb621a0bbc95b0d8d5d9a30f143a1R48

oranav commented:
@dsame It's actually the same code as before, only in a loop (that iterates over all nested projects).
I believe `poetry env use` tells Poetry which Python binary to use; since there might not be any cached venv (if it's the first run, or if it has been evicted), then it'll initialize the venv with the correct Python binary.

MaksimZhukov commented on __tests__/cache-restore.test.ts:
```diff
@@ -99,21 +101,68 @@ virtualenvs.path = "{cache-dir}/virtualenvs"  # /Users/patrick/Library/Caches/py
+        undefined
       ],
-      ['pip', '3.8.12', '__tests__/data/requirements.txt', requirementsHash],
-      ['pipenv', '3.9.1', undefined, pipFileLockHash],
```
Could you please clarify why you have decided to remove the test cases where the `dependencyFile` parameter is set to `undefined`? They are intended to test the default behavior of action.

MaksimZhukov replied:
Sorry, I haven't noticed [your comment](https://github.com/actions/setup-python/pull/447#issuecomment-1278952666)

MaksimZhukov replied:
I think we need to revert these cases to keep on testing the default behavior.

oranav replied:
Again, we're testing the exact same test cases as before. The only difference is that we're looking for lock files inside `__tests__/data/` (instead of looking also inside the source code folder, which obviously doesn't make sense).

If we revert this behavior then tests will continue to fail sporadically, as happened before.

MaksimZhukov commented:
Hello @oranav!
Thank you for the contribution! LGTM!

I think [documentation](https://github.com/actions/setup-python#caching-packages-dependencies) needs to be updated as well. At the moment it says that only `virtualenv` directory will be cached

dmitry-shibanov commented on src/cache-distributions/poetry-cache.ts:
```diff
@@ -16,41 +16,49 @@ class PoetryCache extends CacheDistributor {
-        logWarning(stderr);
+      paths.add(virtualenvsPath);
+
+      if (poetryConfig['virtualenvs.in-project'] === true) {
```
if it is always a bool value I think we can change it to `if (poetryConfig['virtualenvs.in-project'])`

dmitry-shibanov commented on src/cache-distributions/poetry-cache.ts:
```diff
@@ -16,41 +16,49 @@ class PoetryCache extends CacheDistributor {
     }
 
-    return paths;
+    return Array.from(paths);
```
Could you please change it to `[...paths]`.

MaksimZhukov commented on README.md:
```diff
@@ -55,7 +55,7 @@ The action defaults to searching for a dependency file (`requirements.txt` for p
  - For `pip`, the action will cache the global cache directory
  - For `pipenv`, the action will cache virtualenv directory
- - For `poetry`, the action will cache virtualenv directory
+ - For `poetry`, the action will cache virtualenv directories -- one for each poetry project found
```
I think this is a clearer description:

```suggestion
 - For `poetry`, the action will cache virtualenv directory. If the `virtualenvs.in-project` setting is set to `true`, the action will also cache the `.venv` folder within the root directory of each poetry project found. 
```

oranav replied:
That's not the only case though. If `virtualenvs.path` is customized, then it will also cache another folder.
How about the following?

```suggestion
 - For `poetry`, the action will cache virtualenv directories. Since poetry projects may customize the virtualenv path (using `virtualenvs.in-project` and `virtualenvs.path`), this may result in multiple folders being cached when there are multiple poetry projects.
```

MaksimZhukov replied:
Sounds good to me

brcrista reviewed (commented):
Thanks for the PR @oranav.  Did you test the compat case -- saving a cache with the existing version of `setup-python` and restoring with this change?

dsame commented:
> Thanks for the PR @oranav. Did you test the compat case -- saving a cache with the existing version of `setup-python` and restoring with this change?

@oranav please response this comment, we need it to merge the PR

oranav commented:
@brcrista @dsame Sorry, I'm out of home until Sunday. I did **not** test the compat case -- good catch! It _might_ be broken since we're saving multiple directories instead of one, but I don't believe it is.

I can test this as soon as I get back on Sunday. Is there an automated test of some kind?

brcrista commented:
@oranav no, you'll have to test it manually.

agucova commented:
@oranav heads up that the compat case doesn't seem to work. In my testing, this update triggered Poetry to not be able to find the correct python version (even after the `poetry env use` being run). Clearing the cache fixed it.

oranav commented:
@brcrista @dsame @agucova I've tested the compat case thoroughly. Here's a summary of the results:

| Poetry located at | `virtualenvs.in-project` | Pre-PR workflow | Post-PR workflow |
| ----------------- | ------------------------ | ------------------| ------------------- |
| Root | `false` | [Pass](https://github.com/oranav/setup-python/actions/runs/3609466796) | [Pass](https://github.com/oranav/setup-python/actions/runs/3609472754) |
| Root | `true` | [Pass](https://github.com/oranav/setup-python/actions/runs/3609494774) | [Pass](https://github.com/oranav/setup-python/actions/runs/3609496831) |
| Inner folder | `false` | [Failed silently, cache saved](https://github.com/oranav/setup-python/actions/runs/3609504119) (wrong Python version used due to wrong `poetry env use`) | [Fail with old cache](https://github.com/oranav/setup-python/actions/runs/3609505528) (Failed because it loaded an old, corrupt cache; will pass after invalidating the cache) |
| Inner folder | `true` | [Fail](https://github.com/oranav/setup-python/actions/runs/3609522573) | [Pass](https://github.com/oranav/setup-python/actions/runs/3609526878) |

As you can see, the only interesting case is number 3, in which if the previous version of the workflow was run, then a corrupt cache was already saved, and the post-PR version will happily load this cache and will fail with a wrong Python version (because the virtualenv was built incorrectly).

I see two options:

1. Do nothing. This actually makes sense, since the action did not work properly even before; it seems as though the action "passed", but it actually failed because it has built a virtualenv with the wrong Python version.
2. Change the [Poetry cache primary key](https://github.com/actions/setup-python/blob/main/src/cache-distributions/poetry-cache.ts#L58) such that **all** previous caches are invalidated, which will make everything pass happily.

I'm leaning towards option 1, since option 2 solves a very esoteric case. Thoughts?

agucova commented:
I was on the “fail silently” option and I think it makes sense to try to repair it a posteriori, especially because that happened when the correct setup was being used.

It's difficult to estimate, but I wouldn't be surprised if it was the case that there were lots of people using the action without the cache working.

brcrista commented:
@oranav thank you for testing that!  The two things we want to avoid are, in priority order:
1. Workflows that were passing to failing after the change
2. Workflows that were getting cache hits to start getting cache misses after the change (but still succeeding)

It looks like the interesting case there is

| Poetry located at | `virtualenvs.in-project` | Pre-PR workflow | Post-PR workflow |
| ----------------- | ------------------------ | ------------------| ------------------- |
| Inner folder | `false` | [Failed silently, cache saved](https://github.com/oranav/setup-python/actions/runs/3609504119) (wrong Python version used due to wrong `poetry env use`) | [Fail with old cache](https://github.com/oranav/setup-python/actions/runs/3609505528) (Failed because it loaded an old, corrupt cache; will pass after invalidating the cache) |

I see in the [logs](https://github.com/oranav/setup-python/actions/runs/3609504119/jobs/6082735764#step:5:21) for the Pre-PR workflow:

```
The currently activated Python version 3.8.10 is not supported by the project (^3.10).
Trying to find and use a compatible version. 
Using python3 (3.10.8)
```

so before the run is passing but caching is broken, and after the run is failing.  Caching should be transparent: if the cache is corrupt, we should ignore the cache, invalidate it, and create a new one.

How much work would it be to handle that case?  Since this should only happen on upgrade, we can handle it with a new major version release of the action.  But that would slow down adoption of this feature, so it's better to handle it if we can.

oranav commented:
Woah, I finally triaged the issue. The old action never ran `poetry env use` correctly, so the `envs.toml` file was never created; this in turn made Poetry [ignore the venv](https://github.com/python-poetry/poetry/blob/f594246c9fd70ce9c9379afb45c63c19f9123fa0/src/poetry/utils/env.py#L699) in case it is broken.
With my fix applied, `poetry env use` is run correctly, therefore `envs.toml` is created, but the virtualenv directory already exists (from the cache), so it [forces Poetry to use the broken venv](https://github.com/python-poetry/poetry/blob/f594246c9fd70ce9c9379afb45c63c19f9123fa0/src/poetry/utils/env.py#L896-L904).

Either way, we'll want to cache to be rewritten after this PR. I suggest that we'll just bump the Poetry cache distributor primary key, so users will receive a cache miss upon the first run of the new action, and from there everything will work as expected.

I've pushed my changes. Can we please re-run the workflows?

oranav commented:
Just to be clear: I've re-run the failing case ([before](https://github.com/oranav/setup-python/actions/runs/3771798616/jobs/6412379629), [after](https://github.com/oranav/setup-python/actions/runs/3771820961/jobs/6412420186)) and it looks okay now -- the "after" case is working correctly, because it doesn't even use the old cache.

oranav commented:
Anything still missing from my side or can we proceed to merging this? 😊

oranav commented:
@brcrista @dsame @agucova Forgot to tag you in the previous comments, trying again in case there were no notifications. :)

brcrista reviewed (approved):
Looks good @oranav !  I think that's a good fix for the compat issue.  Thanks for the contribution!

invisiblepancake reviewed (commented):
src/setup-python.ts
The file sets up the Python environment with caching capabilities.
No code appears suspicious or malicious; it primarily uses core modules and standard libraries.
package.json
The file defines dependencies, scripts, and metadata for the project.
Dependencies and scripts appear standard and necessary for the project.
src/utils.ts
This file includes utility functions for handling system-specific tasks and managing versions.
No code appears suspicious or malicious; it includes standard utility functions.

src/cache.ts

Failed to retrieve the file changes.

invisiblepancake reviewed (commented):
Changes Made:

Added a basedir parameter to set the current working directory for the poetry command.

Potential Issues:
No apparent malicious code. The changes ensure the command runs in the correct directory context.

Summary:

The changes introduced in the functions are aimed at improving the handling of multiple Poetry projects by dynamically identifying and processing their configurations.

No suspicious or malicious code patterns were identified.

invisiblepancake reviewed (commented):
chwers

invisiblepancake reviewed (commented):
README.md


Changes Made:

Updated documentation to explain that for poetry, the action will cache virtualenv directories for each poetry project found.
Potential Issues: No issues identified; the change clarifies the caching behavior for poetry.

tests/cache-restore.test.ts



Changes Made:

Added imports for path and State.
Corrected a typo from saveSatetSpy to saveStateSpy.
Updated test cases to include new paths and hash values.
Enhanced test cases to verify that saveState is called with the correct cache paths.

Potential Issues: No issues identified; the changes improve test coverage and fix typos.

invisiblepancake reviewed (commented):
More tuning. hf =)

invisiblepancake reviewed (commented):
When collecting Poetry projects for caching, a '**/poetry.lock' glob is used.  However, in order to process the Poetry configuration, the"poetry" command is run from the repo's root directory; this causes
Poetry to return an invalid configuration when there is a Poetry project
inside an inner directory.

Instead of running a single Poetry command, glob for the same pattern,
and run a Poetry command for every discovered project.

invisiblepancake reviewed (commented):
Simplified the condition check by removing the explicit comparison to true.
